src/main/java/hudson/remoting/Channel.java

```diff
@@ -1484,4 +1516,77 @@ public static Channel current() {
         // to avoid situations like this, create proxy classes that we need during the classloading
         jarLoaderProxy=RemoteInvocationHandler.getProxyClass(JarLoader.class);
     }
+
+    private static class ThreadLastIoId extends ThreadLocal<int[]> {
+        @Override
+        protected int[] initialValue() {
+            return new int[1];
+        }
+    }
+
+    /**
+     * A reference for the {@link Channel} that can be cleared out on {@link #close()}/{@link #terminate(IOException)}.
+     * @since FIXME after merge
+     * @see #reference
+     */
+    /*package*/ static final class Ref {
```

Comment:
BTW this whole class could probably be replaced with `AtomicReference<Channel>`, at the loss of a little clarity in naming.